File: bindings/python/flowd/contrib/anthropic.py
"""A ready-made Claude provider for Flow's ``model`` tools.

``effect model("X")`` in a flow declares that a tool is implemented by a model
named ``X``. This provider serves every such tool against Claude with **no host
code**: it derives the structured-output schema from the tool's declared output
type (so the model can't return a malformed value), reports token usage into
the trace, and maps each logical model id to a real Claude model.

    from flowd.contrib.anthropic import anthropic_provider

    rt.register_provider(anthropic_provider(ir))            # one line
    rt.register_provider(anthropic_provider(
        ir,
        model="claude-opus-4-8",
        instructions={"claude-underwriter-v1": "You are a conservative underwriter."},
    ))

Needs the ``anthropic`` SDK (``pip install "flowd[anthropic]"``).
"""

from __future__ import annotations

import json
from typing import Any, Optional

from ..schema import _normalize_ir, json_schema_for


def anthropic_provider(
    ir: Any,
    *,
    model: str | dict = "claude-opus-4-8",
    instructions: Optional[dict] = None,
    thinking: Optional[str] = "adaptive",
    max_tokens: int = 4096,
    client: Any = None,
    name: str = "anthropic",
):
    """Return a provider adapter (for ``Runtime.register_provider``) that serves
    every ``model``-effect tool in ``ir`` against Claude.

    ``model`` is the real Claude model for all served ids, or a
    ``{logical_id: real_model}`` mapping. ``instructions`` optionally overrides
    the system prompt per logical id. ``client`` injects an Anthropic client
    (defaults to ``anthropic.Anthropic()``); pass a stub to test offline.
    """
    doc = _normalize_ir(ir)
    served: dict[str, dict] = {}
    for tool in doc.get("tools", []):
        effect = tool.get("effect", {})
        if effect.get("level") == "model" and effect.get("model"):
            served.setdefault(effect["model"], tool)  # first tool wins per id
    if not served:
        raise ValueError("the IR declares no model() tools for Claude to serve")

    if client is None:
        import anthropic  # lazy: only needed when actually invoking

        client = anthropic.Anthropic()
    instructions = instructions or {}

    def _real_model(model_id: str) -> str:
        return model[model_id] if isinstance(model, dict) else model

    class _AnthropicProvider:
        def __init__(self):
            self.name = name

        def supports_model(self, model_id: str) -> bool:
            return model_id in served

        def _ask(self, model_id: str, request_json: str):
            tool = served[model_id]
            schema = json_schema_for(doc, tool["output"])
            system = instructions.get(model_id) or (
                f"You implement the tool '{tool['name']}'. Given the input, "
                f"return JSON matching the required output schema. Do not explain."
            )
            kwargs = dict(
                model=_real_model(model_id),
                max_tokens=max_tokens,
                system=system,
                messages=[{"role": "user", "content": request_json}],
                output_config={"format": {"type": "json_schema", "schema": schema}},
            )
            if thinking:
                kwargs["thinking"] = {"type": thinking}
            resp = client.messages.create(**kwargs)
            text = next(b.text for b in resp.content if b.type == "text")
            return json.loads(text), getattr(resp, "usage", None)

        def invoke(self, model_id: str, request_json: str):
            value, _ = self._ask(model_id, request_json)
            return value

        def invoke_with_metrics(self, model_id: str, request_json: str):
            value, usage = self._ask(model_id, request_json)
            return {
                "response": value,
                "tokens_in": getattr(usage, "input_tokens", 0) if usage else 0,
                "tokens_out": getattr(usage, "output_tokens", 0) if usage else 0,
                "cost_cents": 0.0,
            }

    return _AnthropicProvider()

File: bindings/typescript/src/anthropic.ts
/**
 * A ready-made Claude provider for Flow's `model` tools.
 *
 * `effect model("X")` declares that a tool is implemented by a model named X.
 * This provider serves every such tool against Claude with no host code: it
 * derives the structured-output schema from the tool's declared output type,
 * reports token usage into the trace, and maps each logical model id to a real
 * Claude model.
 *
 *     runtime.registerProvider(anthropicProvider(ir));
 *
 * Because flowd's callbacks are synchronous, the async Anthropic SDK call is
 * bridged through a bundled worker (anthropic-worker.mjs) — an implementation
 * detail the host never sees. Needs `@anthropic-ai/sdk` at runtime.
 */

import { execFileSync } from "node:child_process";
import * as path from "node:path";

import { jsonSchemaFor, normalizeIr, type IrDocument } from "./schema.js";
import type { ProviderAdapter } from "./index.js";

export interface ClaudeRequest {
  model: string;
  system: string;
  schema: object;
  request: string;
  thinking: string | null;
  maxTokens: number;
}
export interface ClaudeResult {
  response: unknown;
  tokensIn?: number;
  tokensOut?: number;
}
/** The Claude call. Default bridges synchronously to the SDK via a subprocess;
 * inject a stub to test offline. */
export type ClaudeCall = (req: ClaudeRequest) => ClaudeResult;

export interface AnthropicProviderOptions {
  /** Real Claude model for all served ids, or a {logicalId: realModel} map. */
  model?: string | Record<string, string>;
  /** Optional system prompt per logical model id. */
  instructions?: Record<string, string>;
  thinking?: "adaptive" | null;
  maxTokens?: number;
  name?: string;
  call?: ClaudeCall;
}

// The bundled worker; override with FLOWD_ANTHROPIC_WORKER (custom provider,
// or testing the bridge without the SDK). Resolved per call so the override is
// dynamic.
function workerPath(): string {
  return process.env.FLOWD_ANTHROPIC_WORKER ?? path.join(__dirname, "..", "anthropic-worker.mjs");
}

const subprocessCall: ClaudeCall = (req) => {
  const out = execFileSync(process.execPath, [workerPath()], {
    input: JSON.stringify(req),
    encoding: "utf8",
    maxBuffer: 16 * 1024 * 1024,
  });
  return JSON.parse(out) as ClaudeResult;
};

/** Build a provider adapter that serves every `model`-effect tool in `ir`. */
export function anthropicProvider(
  ir: IrDocument | string,
  opts: AnthropicProviderOptions = {},
): ProviderAdapter {
  const doc = normalizeIr(ir);
  const served = new Map<string, { name: string; output: string }>();
  for (const tool of doc.tools ?? []) {
    if (tool.effect.level === "model" && tool.effect.model && !served.has(tool.effect.model)) {
      served.set(tool.effect.model, { name: tool.name, output: tool.output });
    }
  }
  if (served.size === 0) {
    throw new Error("the IR declares no model() tools for Claude to serve");
  }

  const call = opts.call ?? subprocessCall;
  const realModel = (id: string): string =>
    typeof opts.model === "object" ? opts.model[id] : opts.model ?? "claude-opus-4-8";

  const ask = (id: string, request: string): ClaudeResult => {
    const tool = served.get(id)!;
    const schema = jsonSchemaFor(doc, tool.output);
    const system =
      opts.instructions?.[id] ??
      `You implement the tool '${tool.name}'. Given the input, return JSON ` +
        `matching the required output schema. Do not explain.`;
    return call({
      model: realModel(id),
      system,
      schema,
      request,
      thinking: opts.thinking ?? "adaptive",
      maxTokens: opts.maxTokens ?? 4096,
    });
  };

  return {
    name: opts.name ?? "anthropic",
    supportsModel: (id) => served.has(id),
    invoke: (id, req) => ask(id, req).response,
    invokeWithMetrics: (id, req) => {
      const r = ask(id, req);
      return {
        response: r.response,
        tokensIn: r.tokensIn ?? 0,
        tokensOut: r.tokensOut ?? 0,
        costCents: 0,
      };
    },
  };
}
